File: apps/xmtp_sdk_bindgen/runtime/kotlin/streams/Readers.kt
package uniffi.xmtp_sdk

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.async
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

sealed interface SDKStreamCloseReason {
    data object Closed : SDKStreamCloseReason

    data class Failed(
        val error: Throwable,
    ) : SDKStreamCloseReason
}

private fun <T, R> readerFlow(
    owner: SDKClient,
    open: suspend () -> R,
    next: suspend (R) -> T?,
    end: suspend (R) -> Unit,
    connectionState: (R) -> ConnectionState,
    connectionStateChanged: suspend (R, ConnectionState) -> ConnectionState,
    onClose: ((SDKStreamCloseReason) -> Unit)?,
    onConnectionStateChange: ((ConnectionState?, ConnectionState) -> Unit)?,
): Flow<T> =
    flow {
        // The independent opener can finish after collection is cancelled.
        val openerScope = CoroutineScope(Dispatchers.Default)
        val opening = openerScope.async { open() }
        var reader: R? = null
        var failure: Throwable? = null
        val monitor = CoroutineScope(Dispatchers.Default)
        try {
            val active = opening.await()
            reader = active
            if (onConnectionStateChange != null) {
                monitor.launch {
                    var previous: ConnectionState? = null

                    fun emitState(current: ConnectionState) {
                        if (previous == current) return
                        onConnectionStateChange(previous, current)
                        previous = current
                    }
                    emitState(ConnectionState.CONNECTING)
                    emitState(connectionState(active))
                    try {
                        while (true) {
                            emitState(connectionStateChanged(active, previous!!))
                        }
                    } catch (_: Throwable) {
                        // The collection reports terminal read errors.
                    }
                }
            }
            while (true) {
                owner.raw.clientKey()
                val value = next(active) ?: break
                emit(value)
            }
        } catch (error: Throwable) {
            failure = error
            throw error
        } finally {
            monitor.cancel()
            val active = reader
            if (active == null) {
                // Collection can settle before an opener returns. Close its late reader.
                openerScope.launch { runCatching { end(opening.await()) } }
            } else {
                withContext(NonCancellable) {
                    runCatching { end(active) }
                }
            }
            val reason =
                failure
                    ?.takeUnless { it is CancellationException }
                    ?.let(SDKStreamCloseReason::Failed) ?: SDKStreamCloseReason.Closed
            onClose?.invoke(reason)
        }
    }

internal fun messageFlow(
    owner: SDKClient,
    open: suspend () -> MessageReader,
    onClose: ((SDKStreamCloseReason) -> Unit)?,
    onConnectionStateChange: ((ConnectionState?, ConnectionState) -> Unit)?,
): Flow<Message> =
    readerFlow(
        owner,
        open,
        next = { it.next() },
        end = { it.end() },
        connectionState = { it.connectionState() },
        connectionStateChanged = { reader, previous -> reader.connectionStateChanged(previous) },
        onClose = onClose,
        onConnectionStateChange = onConnectionStateChange,
    )

internal fun conversationFlow(
    owner: SDKClient,
    open: suspend () -> ConversationReader,
    onClose: ((SDKStreamCloseReason) -> Unit)?,
    onConnectionStateChange: ((ConnectionState?, ConnectionState) -> Unit)?,
): Flow<Conversation> =
    readerFlow(
        owner,
        open,
        next = { it.next() },
        end = { it.end() },
        connectionState = { it.connectionState() },
        connectionStateChanged = { reader, previous -> reader.connectionStateChanged(previous) },
        onClose = onClose,
        onConnectionStateChange = onConnectionStateChange,
    )

File: apps/xmtp_sdk_bindgen/runtime/swift/streams/Readers.swift
import Foundation

public typealias SDKMessageStream = AsyncThrowingStream<Message, Error>
public typealias SDKConversationStream = AsyncThrowingStream<Conversation, Error>

public enum SDKStreamCloseReason: @unchecked Sendable {
    case closed
    case failed(Error)
}

private extension AsyncThrowingStream where Failure == Error {
    init(
        unfolding: @escaping @Sendable () async throws -> Element?,
        onCancel: @escaping @Sendable () -> Void
    ) {
        self.init(unfolding: {
            try await withTaskCancellationHandler(operation: {
                try await unfolding()
            }, onCancel: onCancel)
        })
    }
}

private final class StreamCompletion: @unchecked Sendable {
    private let lock = NSLock()
    private var didClose = false
    private let callback: (@Sendable (SDKStreamCloseReason) -> Void)?

    init(_ callback: (@Sendable (SDKStreamCloseReason) -> Void)?) {
        self.callback = callback
    }

    func close(_ reason: SDKStreamCloseReason) {
        lock.lock()
        let notify = !didClose
        didClose = true
        lock.unlock()
        if notify {
            callback?(reason)
        }
    }

    var closed: Bool {
        lock.lock()
        defer { lock.unlock() }
        return didClose
    }
}

private func makeStream<Value>(
    owner: SDKClient,
    next: @escaping @Sendable () async throws -> Value?,
    end: @escaping @Sendable () async -> Void,
    connectionState: @escaping @Sendable () -> ConnectionState,
    connectionStateChanged: @escaping @Sendable (ConnectionState) async throws -> ConnectionState,
    onClose: (@Sendable (SDKStreamCloseReason) -> Void)?,
    onConnectionStateChange: (@Sendable (ConnectionState?, ConnectionState) -> Void)?
) -> AsyncThrowingStream<Value, Error> {
    let completion = StreamCompletion(onClose)
    let monitor = Task {
        guard let onConnectionStateChange else { return }
        var previous: ConnectionState?
        func emit(_ current: ConnectionState) {
            guard !completion.closed, previous != current else { return }
            onConnectionStateChange(previous, current)
            previous = current
        }
        emit(.connecting)
        emit(connectionState())
        do {
            while !completion.closed, let last = previous {
                try emit(await connectionStateChanged(last))
            }
        } catch {
            // The pending read reports terminal failure.
        }
    }
    return AsyncThrowingStream<Value, Error>(unfolding: {
        _ = owner.raw
        do {
            let value = try await next()
            if value == nil {
                completion.close(.closed)
                monitor.cancel()
                await end()
            }
            return value
        } catch {
            if !completion.closed {
                completion.close(.failed(error))
            }
            monitor.cancel()
            await end()
            throw error
        }
    }, onCancel: {
        completion.close(.closed)
        monitor.cancel()
        Task { await end() }
    })
}

func makeSDKMessageStream(
    reader: MessageReader,
    owner: SDKClient,
    onClose: (@Sendable (SDKStreamCloseReason) -> Void)?,
    onConnectionStateChange: (@Sendable (ConnectionState?, ConnectionState) -> Void)?
) -> SDKMessageStream {
    makeStream(
        owner: owner,
        next: { try await reader.next() },
        end: { try? await reader.end() },
        connectionState: { reader.connectionState() },
        connectionStateChanged: { try await reader.connectionStateChanged(previous: $0) },
        onClose: onClose,
        onConnectionStateChange: onConnectionStateChange
    )
}

func makeSDKConversationStream(
    reader: ConversationReader,
    owner: SDKClient,
    onClose: (@Sendable (SDKStreamCloseReason) -> Void)?,
    onConnectionStateChange: (@Sendable (ConnectionState?, ConnectionState) -> Void)?
) -> SDKConversationStream {
    makeStream(
        owner: owner,
        next: { try await reader.next() },
        end: { try? await reader.end() },
        connectionState: { reader.connectionState() },
        connectionStateChanged: { try await reader.connectionStateChanged(previous: $0) },
        onClose: onClose,
        onConnectionStateChange: onConnectionStateChange
    )
}
